src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -680,23 +686,153 @@ public new CSharpCompilation AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
-                    builder.Add(tree);
-                    ordinalMap = ordinalMap.Add(tree, oldTreeCount + i);
+                    AppendAllSyntaxTrees(treesBuilder, tree, ordinalMapBuilder, ref loadDirectiveMap, ref declMap, ref declTable, ref referenceDirectivesChanged);
 
                     i++;
                 }
 
-                if (IsSubmission && declMap.Count > 1)
+                return UpdateSyntaxTrees(treesBuilder.ToImmutable(), ordinalMapBuilder.ToImmutable(), loadDirectiveMap, declMap, declTable, referenceDirectivesChanged);
+            }
+            finally
+            {
+                treesBuilder.Free();
+            }
+        }
+
+        /// <summary>
+        /// Appends all tree and any trees from #load'ed files.
+        /// </summary>
+        private void AppendAllSyntaxTrees(
+            ArrayBuilder<SyntaxTree> treesBuilder,
+            SyntaxTree tree,
+            IDictionary<SyntaxTree, int> ordinalMapBuilder,
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
+            ref DeclarationTable declTable,
+            ref bool referenceDirectivesChanged)
+        {
+            // This HashSet will use default (case-sensitive) comparison.
+            // It is the responsibility of the SourceReferenceResolver to
+            // normalize the paths it resolves in a way that is appropriate
+            // for the patforms the host supports.
+            var seenFiles = PooledHashSet<string>.GetInstance();
+            AppendSyntaxTreesHelper(
+                treesBuilder,
+                tree,
+                tree.Options, // Use ParseOptions from external SyntaxTree for all sources it #loads.
+                this.Options,
+                seenFiles,
+                this.MessageProvider,
+                IsSubmission,
+                ordinalMapBuilder,
+                ref loadDirectiveMap,
+                ref declMap,
+                ref declTable,
+                ref referenceDirectivesChanged);
+            seenFiles.Free();
+        }
+
+        /// <summary>
+        /// Recursively adds trees #load'ed by <paramref name="tree"/> to the <paramref name="treesBuilder"/>.
+        /// </summary>
+        private static void AppendSyntaxTreesHelper(
+            ArrayBuilder<SyntaxTree> treesBuilder,
+            SyntaxTree tree,
+            ParseOptions parseOptions,
+            CSharpCompilationOptions options,
+            HashSet<string> seenFiles,
+            CommonMessageProvider messageProvider,
+            bool isSubmission,
+            IDictionary<SyntaxTree, int> ordinalMapBuilder,
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
```

Comment:
They could both be IDictionary, but I wasn't as worried about using a builder for ```loadDirectiveMap``` because we won't likely add to it that much (whereas ```ordinalMapBuilder``` is added to for every ```SyntaxTree```).